cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaVersion.java

```diff
@@ -4,76 +4,68 @@
-        "(?<default>default)?\\s+" +
-        "(?<proto>[0-9.]+)\\s+" +
-        "(?<msg>[0-9.]+)\\s+" +
-        "(?<sha>[0-9A-Za-z]+)\\s+" +
-        "(?<thirdpartylibs>[0-9A-Za-z._-]+)");
-
     /**
      * Parse the version information present in the {@code /kafka-versions} classpath resource.
-     * @param reader A leader for the version info.
+     * @param reader A reader for the kafka version file.
      * @param mapOfVersions A map of the versions to add to.
-     * @return The default version.
-     * @throws Exception
+     * @return The configured default version.
+     * @throws IllegalArgumentException If there are duplicate versions listed in the versions file.
+     * @throws RuntimeException If no default version was set.
+     * @throws IOException If the kafka versions file cannot be read.
      */
-    static KafkaVersion parseKafkaVersions(LineNumberReader reader, Map<String, KafkaVersion> mapOfVersions)
-            throws Exception {
+    public static KafkaVersion parseKafkaVersions(Reader reader, Map<String, KafkaVersion> mapOfVersions) throws IOException {
+
+        YAMLMapper mapper = new YAMLMapper();
+
+        CollectionType listType = mapper.getTypeFactory().constructCollectionType(ArrayList.class, KafkaVersion.class);
+
+        List<KafkaVersion> kafkaVersions = mapper.readValue(reader, listType);
+
         KafkaVersion defaultVersion = null;
-        String line = reader.readLine();
-        while (line != null) {
-            if (!line.isEmpty() && !line.startsWith("#")) {
-                Matcher matcher = KAFKA_VERSION_PATTERN.matcher(line);
-                if (matcher.matches()) {
-                    String version = matcher.group("version");
-                    KafkaVersion kafkaVersion = new KafkaVersion(version,
-                            matcher.group("proto"),
-                            matcher.group("msg"));
-                    if (mapOfVersions.put(version, kafkaVersion) != null) {
-                        throw new Exception("Duplicate version '" + version + "' on line " + reader.getLineNumber());
-                    }
-                    if (matcher.group("default") != null) {
-                        if (defaultVersion == null) {
-                            defaultVersion = kafkaVersion;
-                        } else {
-                            throw new Exception("Multiple default versions given");
-                        }
-                    }
-                } else {
-                    throw new Exception("Malformed line: " + reader.getLineNumber());
-                }
+
+        for (KafkaVersion kafkaVersion: kafkaVersions) {
+
+            if (mapOfVersions.put(kafkaVersion.version, kafkaVersion) != null) {
+                throw new IllegalArgumentException("Duplicate version (" + kafkaVersion.version + ") listed in kafka-versions file");
+            }
+
+            if (kafkaVersion.isDefault) {
+                defaultVersion = kafkaVersion;
```

Comment:
I had an experience of two versions set as default so having the CO failing to start, maybe it could be useful to check that there is no more than one default.

Reply:
I added a check for this which throws an illegal argument exception, do you think it would be worth making custom exceptions to make it more clear where the errors come from?

Reply:
That's a build time error which users should in theory never see, so there's no value in a exception type for that, imho.

Reply:
Agree